src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -597,32 +598,60 @@ private void ResendMessageImpl(Message message, ActivationAddress forwardingAddr
         #region Send path
 
         /// <summary>
-        /// Send an outgoing message
+        /// Send an outgoing message, may complete synchronously
         /// - may buffer for transaction completion / commit if it ends a transaction
         /// - choose target placement address, maintaining send order
         /// - add ordering info and maintain send order
         /// 
         /// </summary>
         /// <param name="message"></param>
         /// <param name="sendingActivation"></param>
-        public async Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
+        public Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
         {
-            try
-            {
-                await AddressMessage(message);
-                TransportMessage(message);
-            }
-            catch (Exception ex)
+            Action<Exception> onAddressingFailure = ex =>
             {
                 if (ShouldLogError(ex))
                 {
-                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed,
-                        String.Format("SelectTarget failed with {0}", ex.Message),
-                        ex);
+                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed, $"SelectTarget failed with {ex.Message}", ex);
                 }
+
                 MessagingProcessingStatisticsGroup.OnDispatcherMessageProcessedError(message, "SelectTarget failed");
                 RejectMessage(message, Message.RejectionTypes.Unrecoverable, ex);
+            };
+
+            Func<Task, Task> transportMessageAfterAddressing = async addressMessageTask =>
```

Comment:
I think that if we completely rule out incorporating this improvement in 1.5 final (most likely anyway), then it would be fine to do that.
Still, I wouldn't hold off from merging this because of that, it could be done separately.